fdbserver/tester.actor.cpp

```diff
@@ -271,104 +272,189 @@ Standalone<VectorRef<KeyValueRef>> checkAllOptionsConsumed(VectorRef<KeyValueRef
+	std::vector<Future<bool>> all;
+	all.reserve(workloads.size());
+	auto wCount = std::make_shared<unsigned>(0);
+	for (int i = 0; i < workloads.size(); i++) {
+		++(*wCount);
+		std::string workloadName = workloads[i]->description();
+		TraceEvent("WorkloadCheckStatus")
+		    .detail("Name", workloadName)
+		    .detail("Count", *wCount)
+		    .detail("Phase", "Start");
+		all.push_back(fmap(
+		    [workloadName, wCount](bool ret) {
+			    --(*wCount);
+			    TraceEvent("WorkloadCheckStatus")
+			        .detail("Name", workloadName)
+			        .detail("Remaining", *wCount)
+			        .detail("Phase", "End");
+			    return true;
+		    },
+		    workloads[i]->check(cx)));
+	}
+	auto done = allTrue(all);
+	if (failureInjection.empty()) {
+		return done;
+	}
+	std::vector<Future<bool>> res;
+	res.reserve(failureInjection.size());
+	for (auto& f : failureInjection) {
+		res.push_back(f->checkInjectionWorkload(cx, done));
+	}
+	return allTrue(res);
+}
+
+ACTOR Future<std::vector<PerfMetric>> getMetricsCompoundWorkload(CompoundWorkload* self) {
+	state std::vector<Future<std::vector<PerfMetric>>> results;
+	for (int w = 0; w < self->workloads.size(); w++) {
+		std::vector<PerfMetric> p;
+		results.push_back(self->workloads[w]->getMetrics());
+	}
+	wait(waitForAll(results));
+	std::vector<PerfMetric> res;
+	for (int i = 0; i < results.size(); ++i) {
+		auto const& p = results[i].get();
+		for (auto const& m : p) {
+			res.push_back(m.withPrefix(self->workloads[i]->description() + "."));
 		}
-		return res;
 	}
+	return res;
+}
 
-	Future<std::vector<PerfMetric>> getMetrics() override { return getMetrics(this); }
-	double getCheckTimeout() const override {
-		double m = 0;
-		for (int w = 0; w < workloads.size(); w++)
-			m = std::max(workloads[w]->getCheckTimeout(), m);
-		return m;
+void CompoundWorkload::addFailureInjection(WorkloadRequest& work) {
+	if (!work.runFailureWorkloads || !FLOW_KNOBS->ENABLE_SIMULATION_IMPROVEMENTS) {
+		return;
+	}
+	// Some common workloads won't work with failure injection workloads
+	for (auto const& w : workloads) {
+		auto desc = w->description();
+		if (desc == "ChangeConfig") {
+			return;
+		} else if (desc == "SaveAndKill") {
+			return;
+		}
+	}
+	auto& factories = IFailureInjectorFactory::factories();
+	DeterministicRandom random(sharedRandomNumber);
+	for (auto& factory : factories) {
+		auto workload = factory->create(*this);
+		while (workload->add(random, work, *this)) {
+			failureInjection.push_back(workload);
+			workload = factory->create(*this);
+		}
 	}
+}
 
-	void getMetrics(std::vector<PerfMetric>&) override { ASSERT(false); }
-};
+Future<std::vector<PerfMetric>> CompoundWorkload::getMetrics() {
+	return getMetricsCompoundWorkload(this);
+}
+
+double CompoundWorkload::getCheckTimeout() const {
+	double m = 0;
+	for (int w = 0; w < workloads.size(); w++)
+		m = std::max(workloads[w]->getCheckTimeout(), m);
+	return m;
+}
+
+void CompoundWorkload::getMetrics(std::vector<PerfMetric>&) {
```

Comment:
This method was introduced for client side workloads. The tester will never call this directly -- calling this would be a bug. So I would rather have this in the C++ file (just to reduce the size of our header files)